quic/s2n-quic-transport/src/ack/ack_ranges.rs

```diff
@@ -0,0 +1,259 @@
+        //             ^--1_2--^
+        //     ^--0_1--^
+        let pn_a = packet_numbers.next().unwrap();
+        let pn_b = packet_numbers.next().unwrap();
+        let pn_c = packet_numbers.next().unwrap();
+        let pn_d = packet_numbers.next().unwrap();
+        let pn_e = packet_numbers.next().unwrap();
+        let pn_f = packet_numbers.next().unwrap();
+        let pn_g = packet_numbers.next().unwrap();
+        let pn_h = packet_numbers.next().unwrap();
+        let pn_i = packet_numbers.next().unwrap();
+        let range_0 = PacketNumberRange::new(pn_a, pn_a);
+        let range_1 = PacketNumberRange::new(pn_b, pn_c);
+        let range_2 = PacketNumberRange::new(pn_d, pn_f);
+        let range_3 = PacketNumberRange::new(pn_g, pn_h);
+        let range_4 = PacketNumberRange::new(pn_h, pn_i);
+        let range_0_1 = PacketNumberRange::new(pn_a, pn_c);
+        let range_1_2 = PacketNumberRange::new(pn_c, pn_e);
+
+        assert!(ack_ranges.insert_packet_number_range(range_1).is_ok());
+        assert!(ack_ranges.insert_packet_number_range(range_2).is_ok());
+        assert!(ack_ranges.insert_packet_number_range(range_3).is_ok());
+        assert_eq!(ack_ranges.interval_len(), 3);
+
+        for pn in range_1 {
+            assert!(ack_ranges.contains(&pn));
+        }
+        for pn in range_2 {
+            assert!(ack_ranges.contains(&pn));
+        }
+        for pn in range_3 {
+            assert!(ack_ranges.contains(&pn));
+        }
+
+        // merge ranges 1 and 2
+        assert!(ack_ranges.insert_packet_number_range(range_1_2).is_ok());
+        assert_eq!(ack_ranges.interval_len(), 2);
+
+        // insert range 0 at low end
+        assert!(ack_ranges.insert_packet_number_range(range_0).is_ok());
+        assert_eq!(ack_ranges.interval_len(), 3);
+
+        // merge range 0_1 at low end
+        assert!(ack_ranges.insert_packet_number_range(range_0_1).is_ok());
+        assert_eq!(ack_ranges.interval_len(), 2);
+
+        // merge new range at high end
+        assert!(ack_ranges.insert_packet_number_range(range_4).is_ok());
+        assert_eq!(ack_ranges.interval_len(), 2);
+    }
+
+    #[test]
+    fn large_range_test() {
+        let pn_a = PacketNumberSpace::ApplicationData.new_packet_number(VarInt::from_u32(1));
+        let pn_b = PacketNumberSpace::ApplicationData
+            .new_packet_number(VarInt::new(varint::MAX_VARINT_VALUE).unwrap());
+        let mut ack_ranges = AckRanges::new(3);
+
+        let range_1 = PacketNumberRange::new(pn_a, pn_b);
+
+        assert!(ack_ranges.insert_packet_number_range(range_1).is_ok());
+        assert_eq!(ack_ranges.interval_len(), 1);
+    }
+
+    #[test]
+    fn size_of_snapshots() {
+        use core::mem::size_of;
+        use insta::assert_debug_snapshot;
+
+        assert_debug_snapshot!("AckRanges", size_of::<AckRanges>());
+    }
+
+    #[test]
+    fn insert_packet_number_range_fuzz() {
+        let mut ack_ranges = AckRanges::new(10);
```

Comment:
Move this inside the for_each. better to allocate a new value each iteration so we don't persist invalid values between inputs.